codec-redis/src/main/java/io/netty/handler/codec/redis/RedisEncoder.java

```diff
@@ -0,0 +1,189 @@
+            writeArrayHeader(allocator, (ArrayHeaderRedisMessage) msg, out);
+        } else if (msg instanceof ArrayRedisMessage) {
+            writeArrayMessage(allocator, (ArrayRedisMessage) msg, out);
+        } else {
+            throw new CodecException("bad message: " + msg);
+        }
+    }
+
+    private void writeSimpleStringMessage(ByteBufAllocator allocator, SimpleStringRedisMessage msg,
+                                                 List<Object> out) {
+        out.add(allocator.ioBuffer(RedisConstants.TYPE_LENGTH).writeByte(RedisMessageType.SIMPLE_STRING.value()));
+        out.add(ByteBufUtil.encodeString(allocator, CharBuffer.wrap(msg.content()), CharsetUtil.UTF_8));
+        out.add(RedisConstants.eolByteBuf());
+    }
+
+    private void writeErrorMessage(ByteBufAllocator allocator, ErrorRedisMessage msg,
+                                                 List<Object> out) {
+        out.add(allocator.ioBuffer(RedisConstants.TYPE_LENGTH).writeByte(RedisMessageType.ERROR.value()));
+        out.add(ByteBufUtil.encodeString(allocator, CharBuffer.wrap(msg.content()), CharsetUtil.UTF_8));
+        out.add(RedisConstants.eolByteBuf());
+    }
+
+    private void writeIntegerMessage(ByteBufAllocator allocator, IntegerRedisMessage msg, List<Object> out) {
+        ByteBuf buf = allocator.ioBuffer(RedisConstants.TYPE_LENGTH + RedisConstants.LONG_MAX_LENGTH +
+                                         RedisConstants.EOL_LENGTH);
+        buf.writeByte(RedisMessageType.INTEGER.value());
+        buf.writeBytes(numberToByteBuf(msg.value()));
+        buf.writeShort(RedisConstants.EOL_SHORT);
+        out.add(buf);
+    }
+
+    private void writeBulkStringHeader(ByteBufAllocator allocator, BulkStringHeaderRedisMessage msg,
+                                              List<Object> out) {
+        final ByteBuf buf = allocator.ioBuffer(RedisConstants.TYPE_LENGTH +
+                                        (msg.isNull() ? RedisConstants.NULL_LENGTH :
+                                                        RedisConstants.LONG_MAX_LENGTH + RedisConstants.EOL_LENGTH));
+        buf.writeByte(RedisMessageType.BULK_STRING.value());
+        if (msg.isNull()) {
+            buf.writeShort(RedisConstants.NULL_SHORT);
+        } else {
+            buf.writeBytes(numberToByteBuf(msg.bulkStringLength()));
+            buf.writeShort(RedisConstants.EOL_SHORT);
+        }
+        out.add(buf);
+    }
+
+    private void writeBulkStringContent(BulkStringRedisContent msg, List<Object> out) {
+        out.add(msg.content().retain());
+        if (msg instanceof LastBulkStringRedisContent) {
+            out.add(RedisConstants.eolByteBuf());
+        }
+    }
+
+    private void writeFullBulkStringMessage(ByteBufAllocator allocator, FullBulkStringRedisMessage msg,
+                                               List<Object> out) {
+        out.add(allocator.ioBuffer(RedisConstants.TYPE_LENGTH).writeByte(RedisMessageType.BULK_STRING.value()));
```

Comment:
For the `Full` message case we should know the length of the message. Can we just allocate a buffer of the appropriate size and write into it? Best to avoid allocating smaller ByteBufs if possible.